ENH: Add support for dataclasses in the DataFrame constructor

asosnovsky opened the pull request:
Added support for data-classes when used in the construction of a new dataframe.

- [X] closes #21910
- [X] tests added / passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

Simply put, added support to use dataclasses in the following way:

```python
from dataclasses import dataclass

@dataclass 
class Person:
    name: str
    age: int

df = DataFrame([Person("me", 25), Person("you", 35)])
```

asosnovsky committed:
- implemented type-coercion for dataclasses
- added a is_dataclass_inst function to detect dataclasses

asosnovsky committed:
added unit tests

asosnovsky committed:
added github issue to unit test

pep8speaks commented:
Hello @asosnovsky! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pandas/core/dtypes/inference.py`](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py):

> [Line 455:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L455): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 458:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L458): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 462:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L462): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 465:37](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L465): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 466:89](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L466): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 88 characters)
> [Line 483:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L483): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 483:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/core/dtypes/inference.py#L483): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file



* In the file [`pandas/tests/frame/test_constructors.py`](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py):

> [Line 1338:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py#L1338): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1347:1](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py#L1347): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 1348:25](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py#L1348): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1348:37](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py#L1348): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1350:20](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py#L1350): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 1351:20](https://github.com/pandas-dev/pandas/blob/7de46ca74d5239e9c341f71f7013da5ec018c97a/pandas/tests/frame/test_constructors.py#L1351): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

asosnovsky committed:
pep8 compliance

asosnovsky committed:
masking functions for lower versions of python

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -16,6 +16,7 @@
 import sys
 from textwrap import dedent
 from typing import FrozenSet, List, Optional, Set, Tuple, Type, Union
+
```
the linter isn't going to like this.  run `isort -rc pandas` to automatically fix this

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -437,10 +440,12 @@ def __init__(self, data=None, index=None, columns=None, dtype=None, copy=False):
                 data = list(data)
             if len(data) > 0:
-                if is_list_like(data[0]) and getattr(data[0], "ndim", 1) == 1:
+                if (is_list_like(data[0]) and getattr(data[0], "ndim", 1) == 1):
```
i dont think the linter is going to like this.  run `black pandas` to automatically fix

jbrockmendel commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1336,6 +1336,26 @@ def test_constructor_list_of_namedtuples(self):
+        try:
+            from dataclasses import dataclass
+        except ImportError:
+            return;
```
no semicolon

asosnovsky committed:
ran black pandas

asosnovsky committed:
corrected index issue

TomAugspurger commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1336,6 +1336,23 @@ def test_constructor_list_of_namedtuples(self):
+        try:
+            from dataclasses import dataclass
+        except ImportError:
+            return
```
Write this as a `pytest.mark.skipif(not compat.PY36)` instead.

TomAugspurger commented on pandas/core/frame.py:
```diff
@@ -440,6 +443,8 @@ def __init__(self, data=None, index=None, columns=None, dtype=None, copy=False):
             if not isinstance(data, abc.Sequence):
                 data = list(data)
             if len(data) > 0:
+                if is_dataclass_instance(data[0]):
```
We need to document this behavior (that the first item is inspected for a dataclass) somewhere. Probably in `dsintro.rst`.

jorisvandenbossche replied:
It's the same as with dicts, or Series, ... (so would need to check if that is documented, otherwise need to add documentation in general about this).

asosnovsky replied:
where is the proper place to add this to the documentation? this is my first contribution

TomAugspurger replied:
These cases seem to be documented in `getting_started/dsintro.rst`. A short section on "From a list of Dataclasses" would be nice.

asosnovsky replied:
@TomAugspurger I just noticed, there is no entry in that file for NamedTuples. The conversion for namedtuples is similar to what I am doing, the constructor checks the first entry and uses the fields provided by the element as a generalization for the rest. I will update the documentation for the dataclasses sometime tomorrow, but do you think we should open a PR for the namedtuples?

jorisvandenbossche replied:
> and uses the fields provided by the element as a generalization for the rest.

This is something that is actually buggy IMO, and so best not to mention it in the docs. See https://github.com/pandas-dev/pandas/issues/27329 (but a PR to fix this is certainly welcome)

asosnovsky replied:
I will try and see, I think that could be solved by converting all namedtuples to dict as well, that way they could leach off the implementation as well.

TomAugspurger reviewed (commented):
Needs a release note (1.0.rst).

A few edge cases we should have tests for

1. A mixture of dataclass types  (probably takes the union of all the columns?)
2. A mixture of dataclass and non-dataclass (this probably raises. Can we ensure that the error message is reasonable)

asosnovsky committed:
removed py3.7 specific syntax
using make_dataclass in tests

asosnovsky committed:
corrected pytest requirements

asosnovsky committed:
corrected unit tests

asosnovsky committed:
dataclasses require py37 not py36

asosnovsky committed:
added additional unit tests for more specific cases

asosnovsky committed:
removed white space

asosnovsky committed:
docs

TomAugspurger commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,27 @@ The result will be a DataFrame with the same index as the input Series, and
 
+From a list of dataclasses
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
```
```suggestion

.. versionadded:: 1.0.0
```

asosnovsky committed:
Update doc/source/getting_started/dsintro.rst

Co-Authored-By: Tom Augspurger <TomAugspurger@users.noreply.github.com>

jorisvandenbossche commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,28 @@ The result will be a DataFrame with the same index as the input Series, and
+    data = [Point(0,0), HLine(0,3,10), Point(2,3)]
+    df = pd.DataFrame(data)
+
+.. _basics.dataframe.from_list_dataclasses:
```
This label goes above the title

jorisvandenbossche commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,28 @@ The result will be a DataFrame with the same index as the input Series, and
+dataclasses as introduced in `PEP557 <https://www.python.org/dev/peps/pep-0557>`,
+can be passed into the DataFrame constructor.
+Passing a list of dataclasses is equivilent to passing a list of dictionaries. However,
+it is imperative that all values in the list are dataclasses otherwise expect a TypeError.
```
The "however" is not fully correct (as it seems to imply this is not the case for a list of dictionaries; but it is the same for that case)

asosnovsky replied:
@jorisvandenbossche  how about:

Passing a list of dataclasses is equivilent to passing a list of dictionaries. 

Please be aware, that that all values in the list should be dataclasses, mixing types in the list would result in a TypeError.

asosnovsky replied:
I updated the docs with the text above, please let me know if I should change it again

jorisvandenbossche commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,28 @@ The result will be a DataFrame with the same index as the input Series, and
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 1.0.0
+dataclasses as introduced in `PEP557 <https://www.python.org/dev/peps/pep-0557>`,
```
Can you leave a blank line between the `.. versionadded:: ` and this one? (sphinx is rather picky about that ..)

jorisvandenbossche replied:
```suggestion
Data Classes as introduced in `PEP557 <https://www.python.org/dev/peps/pep-0557>`__,
```

jorisvandenbossche commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,28 @@ The result will be a DataFrame with the same index as the input Series, and
+    Point = make_dataclass("Point", [("x", int), ("y", int)])
+    HLine = make_dataclass("HLine", [("x0", int), ("x1", int), ("y", int)])
+    
+    data = [Point(0,0), HLine(0,3,10), Point(2,3)]
```
I would show an example with all the same dataclasses, that is much more typical (we should encourage passing different ones IMO).

jorisvandenbossche reviewed (commented):
Added some comments on the doc addition!

asosnovsky committed:
updated docs

asosnovsky committed:
Merge branch 'issue/21910' of github.com:asosnovsky/pandas into issue/21910

asosnovsky committed:
Update doc/source/getting_started/dsintro.rst

Co-Authored-By: Joris Van den Bossche <jorisvandenbossche@gmail.com>

asosnovsky committed:
Merge branch 'issue/21910' of github.com:asosnovsky/pandas into issue/21910

asosnovsky committed:
updated doc/source/getting_started/dsintro.rst

asosnovsky commented:
> Needs a release note (1.0.rst).
> 
> A few edge cases we should have tests for
> 
>     1. A mixture of dataclass types  (probably takes the union of all the columns?)
> 
>     2. A mixture of dataclass and non-dataclass (this probably raises. Can we ensure that the error message is reasonable)

0. I updated the unit-tests for these cases [see here](https://github.com/asosnovsky/pandas/blob/issue/21910/pandas/tests/frame/test_constructors.py#L1340)

1. since the evaluation of dataclasses simply latches onto that of dictionaries, mixing dataclasses would result in more columns ([see this](https://github.com/asosnovsky/pandas/blob/issue/21910/pandas/tests/frame/test_constructors.py#L1355)).

2. A mixture of none-dataclasses and dataclasses will result in a TypeError [see here](https://github.com/asosnovsky/pandas/blob/issue/21910/pandas/tests/frame/test_constructors.py#L1364) I also added a note in the documentation about this [link](https://github.com/asosnovsky/pandas/blob/issue/21910/doc/source/getting_started/dsintro.rst#from-a-list-of-dataclasses).

asosnovsky committed:
ran linters

asosnovsky committed:
ran isort

TomAugspurger reviewed (commented):
Not entirely sure about the doc build failure yet.

If you merge master and repush, the codecov failure on travis will go away.

asosnovsky committed:
Update dsintro.rst

asosnovsky committed:
correcting doc

asosnovsky committed:
Merge branch 'master' into issue/21910

asosnovsky commented:
> Not entirely sure about the doc build failure yet.
@TomAugspurger 
> 
> If you merge master and repush, the codecov failure on travis will go away.
First of, thank you for the help!! :)

I applied your previous comments and re-merged with master. Now I got only 2 failures, going to take a look at it.

asosnovsky committed:
not overriding the data variable under documentation

TomAugspurger reviewed (approved):
LGTM, though I have a few questions about the conversion.

1. Does `dataclass.asdict` make any guarantees about the order of the dictionary (for Python >3.6?) I don't see anything in the docstring. We don't have any tests that would exercise this (non-sorted keys)
2. It seems that `dataclass.asdict` will recursively convert fields of a few types (data classes, dicts, lists, and tuples). What happens here?

```python
@dataclass
class Point:
     x: int
     y: int

@dataclass
class C:
     mylist: List[Point]

p = Point(10, 20)
assert asdict(p) == {'x': 10, 'y': 20}

c = C([Point(0, 0), Point(10, 4)])
assert asdict(c) == {'mylist': [{'x': 0, 'y': 0}, {'x': 10, 'y': 4}]}
```

In theory, we should be able to handle it well, with the `mylist` field being an object dtype. Can you add a test?

jreback commented:
this needs a perf check to make sure no regressions

WillAyd commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1336,6 +1336,34 @@ def test_constructor_list_of_namedtuples(self):
+        result = DataFrame(datas)
+        tm.assert_frame_equal(result, expected)
+
+        # varying types
```
Could you make this a separate test?

asosnovsky replied:
I split the tests into 3 different tests

WillAyd commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,28 @@ The result will be a DataFrame with the same index as the input Series, and
+can be passed into the DataFrame constructor.
+Passing a list of dataclasses is equivilent to passing a list of dictionaries.
+
+Please be aware, that that all values in the list should be dataclasses, mixing
```
Can you add a test that this happens?

WillAyd reviewed (changes requested):
Very nice PR. Just a few comments I think can make the test coverage better / more explicit

WillAyd commented on pandas/tests/frame/test_constructors.py:
```diff
@@ -1336,6 +1336,34 @@ def test_constructor_list_of_namedtuples(self):
+        tm.assert_frame_equal(result, expected)
+
+        # expect TypeError
+        with pytest.raises(TypeError):
```
Sorry didn't notice this tucked away here so ignore comment asking to add, but typically for things that raise we create them as a separate test. So would be better to break this off into a test called `test_constructor_dataclass_with_others_raises` or something to the effect

asosnovsky replied:
I did as asked

TomAugspurger replied:
I think Point isn't defined anymore. Can you move the definition of those to somewhere else? Probably just at the top of this file, not sure.

asosnovsky committed:
split up unit tests

jreback commented on pandas/core/dtypes/inference.py:
```diff
@@ -450,3 +450,38 @@ def is_sequence(obj):
+
+    """
+    try:
+        from dataclasses import is_dataclass
```
I think we can actually put this at the top with a try/except around it

TomAugspurger replied:
That would require a global `HAS_DATCLASSES` variable, which I don't find as clean as this.

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -422,6 +422,12 @@ def _get_axes(N, K, index, columns):
     return index, columns
 
 
+def _dataclasses_to_dicts(data):
```
same here you can put the import at the top with a try/except around it

asosnovsky replied:
I think that is unnecessary as this function should error if someone is calling it in the context where dataclasses are not available. The only thing a try-except add here is ambiguity about the return value (if dataclasses don't exists return empty dict?) and add an additional overhead (try/except is not a cost-free abstraction)

jreback replied:
k

jreback commented on pandas/core/frame.py:
```diff
@@ -440,6 +442,8 @@ def __init__(self, data=None, index=None, columns=None, dtype=None, copy=False):
                 data = list(data)
             if len(data) > 0:
+                if is_dataclass_instance(data[0]):
+                    data = _dataclasses_to_dicts(data)
```
should this not be *after* the is_named_tuples check (below)?

TomAugspurger replied:
That's nested under an `is_list_like(data[0])`. I don't think dataclasses are considered list like.

asosnovsky replied:
correct, a `dataclass` is not "list like" so moving that there would just break this PR.

asosnovsky committed:
brought Point back

jreback commented:
@asosnovsky can you merge master and we'll have another look.

jbrockmendel commented:
@asosnovsky can you rebase and we'll see why the linter failed

WillAyd commented:
@asosnovsky this looks close; logs aren't available from failure but can you merge master and repush to see if that solves?

asosnovsky committed:
added space for formating pass

asosnovsky committed:
merge with current master

asosnovsky committed:
finishing merge

asosnovsky commented:
> @asosnovsky this looks close; logs aren't available from failure but can you merge master and repush to see if that solves?

sorry it took so long, got really busy at work, merged with master and pushed again.

jreback commented:
would be a nice change, can you merge master and ping on green.

asosnovsky committed:
Merge branch 'master' into issue/21910

WillAyd commented on pandas/core/frame.py:
```diff
@@ -462,6 +463,8 @@ def __init__(
             if not isinstance(data, (abc.Sequence, ExtensionArray)):
                 data = list(data)
             if len(data) > 0:
+                if is_dataclass_instance(data[0]):
```
I think the import of this is missing which is causing CI failures

asosnovsky committed:
fixing missing import

asosnovsky committed:
Merge branch 'upstream/master' into issue/21910

asosnovsky committed:
corrected isort

jreback commented on pandas/core/dtypes/common.py:
```diff
@@ -28,6 +28,7 @@
     is_array_like,
     is_bool,
     is_complex,
+    is_dataclass_instance,
```
just call is_dataclass

asosnovsky replied:
a dataclass is a class that is defined with `@dataclass`, I think that `is_dataclass` would imply that we are checking if a class is a dataclass, while this name implies that what I am looking at should be a dataclass instance (not I even have the `not isinstance(item, type)` in the check)

jreback replied:
this is not correct these are all type inferences routines of instances so adding _instance is redundant and not conforming with the style
and is_dataclass is appropriate

asosnovsky replied:
ok

asosnovsky replied:
sorry to circle back on this, but I believe that I am quite right here, given how the documentation refers to dataclasses https://docs.python.org/3/library/dataclasses.html#dataclasses.is_dataclass

The documentation clearly lists a distinction when it says "Return True if its parameter is a dataclass or an instance of one"

The dataclass decorator and the make_dataclass function both return a dataclass object from which dataclass instances can be made.

If I declared

```python
Point = make_dataclass("Point", [("x", int), ("y", int)])
p = Point(0,1)
```

Then `Point` is the dataclass and `p`  is an instance of the dataclass `Point`.

This check looks to see if its the latter and not the former.

jreback commented on pandas/core/dtypes/inference.py:
```diff
@@ -422,3 +422,38 @@ def is_sequence(obj) -> bool:
         return False
+
+
+def is_dataclass_instance(item):
```
item -> obj

jreback commented on pandas/core/dtypes/inference.py:
```diff
@@ -422,3 +422,38 @@ def is_sequence(obj) -> bool:
+    ----------
+    item : object
+
+    Returns
```
add a versionadded tag

asosnovsky replied:
not sure what you mean

WillAyd replied:
If you grep for versionversionadded you will see this in quite a few places. Here's an example:

https://github.com/pandas-dev/pandas/blob/8ff2ebd9b2cfbe4a3e1eb7893dc4343c01ad55b4/pandas/core/indexes/base.py#L1054

Should target 1.1.0 at this point

jreback replied:
can you update

jreback commented on pandas/core/frame.py:
```diff
@@ -110,6 +112,7 @@
 from pandas.core.indexing import check_bool_indexer, convert_to_index_sliceable
 from pandas.core.internals import BlockManager
 from pandas.core.internals.construction import (
+    _dataclasses_to_dicts,
```
de-privatize this

asosnovsky replied:
ok

jreback commented on pandas/core/internals/construction.py:
```diff
@@ -430,6 +430,12 @@ def _get_axes(N, K, index, columns):
     return index, columns
 
 
+def _dataclasses_to_dicts(data):
```
can you try to type things (you can add in a TYPE_CHECKING block if need be)

asosnovsky replied:
ok

jreback reviewed (changes requested):
looks pretty good, some changes, ping on green.

asosnovsky committed:
- renamed is_dataclass_instance to is_dataclass
-  renamed _dataclasses_to_dicts to dataclasses_to_dicts
- added docs to dataclasses_to_dicts

asosnovsky committed:
ran isort and linters

WillAyd commented on doc/source/getting_started/dsintro.rst:
```diff
@@ -397,6 +397,28 @@ The result will be a DataFrame with the same index as the input Series, and
+From a list of dataclasses
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 1.0.0
```
```suggestion
.. versionadded:: 1.1.0
```

jreback replied:
can you update

MarcoGorelli commented:
Hi @asosnovsky  - sorry to chase you up, just wanted to ask whether you're still working on this :)

jreback reviewed (changes requested):
looks good. some minor versionadded docs need updating. pls merge master and ping on green.

MarcoGorelli commented:
Hi @asosnovsky , is this active? I'm hesitant to close is as stale as it looks like it's almost done, but we haven't heard from you in over 2 months

jreback committed:
Update doc/source/getting_started/dsintro.rst

Co-Authored-By: William Ayd <william.ayd@icloud.com>

jreback commented:
this looks fine, but needs a rebase if someone wants to merge master. ping on green.

WillAyd committed:
Merge remote-tracking branch 'upstream/master' into issue/21910

WillAyd committed:
doctest fix

WillAyd committed:
doc fixup

WillAyd commented:
@jreback green

jreback commented:
thanks @asosnovsky

simonjayhawkins commented:
Thanks @asosnovsky Very nice. a few comments to follow-up.

1. I think we should have some mention of this enhancement a whatsnew.
2. #21910 also suggested a to_dataclasses method so maybe should not be closed, or open another issue to discuss this further.
3. Would it make sense to also add dataclass support to the Series constructor for consistency? although the field names would now be the index and dtype would be either object to maintain the field types or maybe coerced?

asosnovsky commented:
> Thanks @asosnovsky Very nice. a few comments to follow-up.
> 
> 
> 
> 1. I think we should have some mention of this enhancement a whatsnew.
> 
> 2. #21910 also suggested a to_dataclasses method so maybe should not be closed, or open another issue to discuss this further.
> 
> 3. Would it make sense to also add dataclass support to the Series constructor for consistency? although the field names would now be the index and dtype would be either object to maintain the field types or maybe coerced?

Not sure if it makes sense to add a "to_dataclasses" method. Would you have to create a dataclass on the fly or feed in one?

Not sure about series either. But if we were to do that, don't series need to have a single dtype? What happens to a dataclass with a mixed bag of types ?

mathbunnyru commented:
I think this feature is quite cool and worth mentioning here:
https://github.com/pandas-dev/pandas/blob/master/doc/source/whatsnew/v1.1.0.rst

jreback commented:
hmm this originally did have a whats new note ; must have gotten lost

would take a patch to add one

mathbunnyru commented:
@asosnovsky please do that. I'm not a native speaker and I think it'll be easier for you to describe the change :)
